Update paho-mqtt-c 1.3.4 conanfile to match CMake options

icraggs opened the pull request:
Specify library name and version:  paho-mqtt-c/1.3.4

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

The options previously used didn't match those in the CMakeLists.txt.

uilianries commented on recipes/paho-mqtt-c/all/conanfile.py:
```diff
@@ -14,11 +14,11 @@ class PahoMqttcConan(ConanFile):
                "fPIC": [True, False],
                "ssl": [True, False],
-               "asynchronous": [True, False]}
+               "samples": [True, False]}
```
why do you need sample, if the idea is only package the library?

icraggs replied:
The samples included the command line utilities that use the library.  Would those not be useful?

uilianries replied:
Yes, they are useful, I remember that I used it. Keep it as optional, but disabled.

icraggs replied:
Samples option included, default False.

icraggs replied:
Samples option included.  Default False.

uilianries commented on recipes/paho-mqtt-c/all/conanfile.py:
```diff
@@ -14,11 +14,12 @@ class PahoMqttcConan(ConanFile):
+    default_options = {"shared": True,
                        "fPIC": True,
-                       "ssl": False,
-                       "asynchronous": True}
```
You can remove it, because it can break user which is using this option explicitly.

If it changed, you remove by config_options, but with a condition based on package version.

icraggs replied:
Actually looking closer, this has not changed in the project.  But the original contribution of this package (not written by me) has a patch which builds either paho-mqtt3cs OR paho-mqtt3as.  Elsewhere I always build both libraries - they are essentially two difference APIs.  One is easier to use, the other more capable.  I'd prefer to include both libraries with every build as is done elsewhere.  What do you think about that?

icraggs replied:
Asynchronous option removed.

uilianries commented on recipes/paho-mqtt-c/all/conanfile.py:
```diff
@@ -14,11 +14,12 @@ class PahoMqttcConan(ConanFile):
                        "fPIC": True,
-                       "ssl": False,
-                       "asynchronous": True}
+                       "ssl": True,
```
When ssl changed? Was it a bug, or is just following the new release?

icraggs replied:
I think it's better that ssl is enabled by default.  I think most people would want to use secure connections.  Is there a reason for not enabling it?

keysight-daryl replied:
totally agree on...
- ssl being enabled by default
- static library by default
- building both asynchronous and classic interface libraries together (even though i think most people use the asynchronous apis) as long as it doesn't cause linker issues or inflate the end executable due to the linker optimizing out all unused code

Croydon commented on recipes/paho-mqtt-c/all/conanfile.py:
```diff
@@ -14,11 +14,12 @@ class PahoMqttcConan(ConanFile):
-    default_options = {"shared": False,
+               "samples": [True, False]}
+    # static builds didn't really work until 1.3.4
+    default_options = {"shared": True,
```
Please don't change this. Statis should be the default for all recipes who support static at all

icraggs replied:
Static didn't work properly until 1.3.4 as that was never the intention when the library was first written in 2008.

Croydon replied:
Okay, but then please only change the default to shared for < 1.3.4

icraggs replied:
I haven't found a way to have version specific defaults.  One way would be to remove the older versions of the library, in due course.

uilianries replied:
What do you mean by don't work properly? Paho's recipe is not new, there is an older recipe in Conan Community and we didn't receive any issue about. Do you have any evidence, like upstream issues?

keysight-daryl replied:
I think he was talking about the windows variant regarding not working correctly as static.  I have been using the library for a couple years now and it is true that while static will indeed pass a build on Windows it won't run properly and I had to force my downstream recipes and packaging to shared for Windows builds.  Ian is the original author of this library so I think it is best we accept his advice on using it and move forward with this much needed set of fixes.

icraggs replied:
That was one reason for creating a separate recipe for 1.3.4 - the default for the shared option could be different for 1.3.4 and the previous versions.  In any case, I have reverted to one recipe now, as requested, and set the default for all for the time being, to shared.

uilianries replied:
Hey Ian, I owe you a beer! I used paho mqtt c for a couple projects in the past!

If it's possible we such configuration (static) we can drop by throwing a ConanInvalidConfiguration and using shared by default. WDYT ?

icraggs replied:
@uilianries Cool!  As we're not likely to meet, if you want to buy me a beer you could sponsor me on Github for a month or two for a couple of $ :-)

Throwing an invalid configuration for windows static builds < 1.3.4 sounds like a good option.  I'll look into it.

I've set the ssl option default to True because just about everyone will want encryption, and I've fixed the SSL pre 1.3.4 builds in a patch.

Also, I think the async builds are more popular than the other variant so that should be the default.  However, the MQTTAsync and MQTTClient libraries are actually different APIs, so I was wondering if they should be in different recipes entirely?

uilianries replied:
Both can be used on same recipe, because we use 1 project x 1 recipe. I understand your point, looking as a developer, I can be interested on only one of them (async, for instance), but in terms of size and option is too small to get it separated.

keysight-daryl replied:
I agree most people probably use the async api and library but a good optimizing linker should optimize out unused functions if linked statically so it probably makes sense to just leave them in the same recipe as long as it doesn't cause conan/cmake confusion in linking against both libs in the same package.  Is it possible to just output a single .lib with both interfaces always might be simpler to deal with in general.

icraggs commented:
I separated the 1.3.4 and previous configurations so that we could concentrate on getting 1.3.4 right.  I intend to remove the superfluous 1.3.4 config from the 'all' folder, but it wouldn't let me do that in one PR.

icraggs commented:
@uilianries It's only duplicated because I couldn't change both recipes in one PR as your process apparently disallows it.  So I was going to remove the duplicate entries from the 'all' folder in another PR to get around this restriction.

I created a new recipe because that seemed the best (only?) way to have different parameter defaults for different versions.  As I changed the defaults for 1.3.4, some of the builds for 1.3.1/1.3.0 failed.  I also figured that maybe the defaults shouldn't be changed for the older versions in case someone was relying on them.

I'll look into changing it back to use the 'all' folder, but on the evidence so far, it may well take me a while to figure out a build that works for all 3 versions.

keysight-daryl commented:
I have built against these proposed changes and using both the paho-c and paho-cpp libraries as shared (had to remove the paho-mqtt-cpp recipe "Paho cpp can not be built as shared on windows" exception in the configure() function) and it seems to fix the original SSL issue I raised in #1882 .  I really wanted to link this statically but if I do that I get a linker error around repeat OpenSSL function definitions existing perhaps due to also linking in Poco statically which uses OpenSSL.  So in the end it seem we will have to use this shared anyway which is fine.

icraggs commented:
I added the configuration exception for windows static builds < 1.3.4.

I added back the asynchronous build option for 1.3.4, and changed it so that the 
- asynchronous
- shared
- ssl
options build one library with just the selected options, as I think they did before.  I think that should be everything.

icraggs commented:
The shared option default is still set to True because of the limitations of the static build on Windows.  As it now works on 1.3.4 then the default could be set to False if you all prefer.

keysight-daryl commented:
> The shared option default is still set to True because of the limitations of the static build on Windows. As it now works on 1.3.4 then the default could be set to False if you all prefer.

I think what @Croydon was saying is the 'standard' is to assume shared as True where possible.  Probably due to the sort of weird difficult linkage issues you get when libraries are linked statically and causing the sort of duplicate definition OpenSSL issues we are seeing in our project.  Static is nice for embedded and optimized binary size but it seems difficult to manage on larger projects in general.

keysight-daryl commented:
Confirming the latest proposed changes are now producing only the single optioned library within the package...
![Screen Shot 2020-06-23 at 1 03 46 PM](https://user-images.githubusercontent.com/40178586/85433348-4c48b300-b552-11ea-9002-e5fe656e037b.png)

keysight-daryl commented:
It appears the latest change to only package one library as optioned also fixes my previous duplicate function linker issue using the library statically.  I am now able to use it statically within our application without the linker throwing it's hands up.  Nice work @icraggs

keysight-daryl commented:
I've noticed in the patched cmakelist it pulls OpenSSL in as 'SHARED' which I was told in the official Cpplang:conan slack channel could cause duplicate symbols issue downstream for what it's worth.
![Screen Shot 2020-06-23 at 2 24 05 PM](https://user-images.githubusercontent.com/40178586/85441593-41475000-b55d-11ea-9994-f2dc1648ffa3.png)

keysight-daryl commented:
Any chance of actually getting this approved and merged?  paho-mqtt-cpp needs associated updates as well

uilianries commented:
it requires more 2 approvals. I've added one more reviewer ...

icraggs commented:
Thanks @uilianries !
